src/js/jsx/search/SearchBar.jsx

```diff
@@ -213,82 +216,131 @@ define(function (require, exports, module) {
+
+            if (categories.size > 0 && this.state.groupedOptions) {
+                optionGroups = this.state.groupedOptions.unshift(categories);
+            } else {
+                return Immutable.List();
+            }
 
-            return options && options.filter(function (option) {
-                if (option.hidden) {
-                    return false;
-                }
+            var filteredOptionGroups = optionGroups.map(function (options) {
+                return options && options.filter(function (option) {
+                    option.priority = 1; // Add to priority to indicate less important
 
-                // Always add headers to list of searchable options
-                // The check to not render if there are no options below it is in Select.jsx
-                if (option.type === "header") {
-                    count = 0;
-                    return true;
-                }
+                    if (option.hidden) {
+                        return false;
+                    }
 
-                if (count === MAX_OPTIONS) {
-                    return false;
-                }
+                    // Always add headers to list of searchable options
+                    // The check to not render if there are no options below it is in Select.jsx
+                    if (option.type === "header") {
+                        count = 0;
+                        option.priority = -1;
+                        return true;
+                    }
 
-                var useTerm = true,
-                    title = option.title.toLowerCase(),
-                    category = option.category || [];
-            
-                // If it is the filter option for something that we already have filtered, don't
-                // show that filter option
-                if (option.id.indexOf("FILTER") === 0 && _.isEqual(this.state.filter, category)) {
-                    return false;
-                }
+                    var useTerm = true,
+                        title = option.title.toLowerCase(),
+                        category = option.category || [];
+                
+                    // If it is the filter option for something that we already have filtered, don't
+                    // show that filter option
+                    if (option.id.indexOf("FILTER") === 0 && _.isEqual(this.state.filter, category)) {
+                        return false;
+                    }
+                    
+                    option.priority++;
+
+                    if (this.state.filter.length > 0) {
+                        // All terms in this.state.filter must be in the option's category
+                        _.forEach(this.state.filter, function (filterValue) {
+                            if (!_.contains(category, filterValue)) {
+                                useTerm = false;
+                            }
+                        });
+
+                        if (!useTerm) {
```

Comment:
It seems like if you ever encounter the line `useTerm = false` above you just want to end this function immediately. If so, you should use `.some` instead of `.forEach` to exit the loop immediately so that you can return immediately.